File: near/src/main/java/com/adroitandroid/near/discovery/client/UdpBroadcastService.java
package com.adroitandroid.near.discovery.client;


import android.app.Service;
import android.content.Intent;
import android.os.IBinder;

public class UdpBroadcastService extends Service {

    public static final String BUNDLE_NAME = "bundle_name";
    public static final String BUNDLE_ACTION = "bundle_action";
    public static final String BUNDLE_INTERVAL = "bundle_interval";
    public static final String ACTION_START_BROADCAST = "start_broadcast";
    public static final String ACTION_STOP_BROADCAST = "stop_broadcast";
    private static final long DEFAULT_BROADCAST_INTERVAL = 5000;
    private BroadcastThread broadcastThread;

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null) {
            if (ACTION_START_BROADCAST.equals(intent.getStringExtra(BUNDLE_ACTION))) {
                if (broadcastThread != null) {
                    broadcastThread.stopBroadcast();
                }
                broadcastThread = new BroadcastThread(intent.getStringExtra(BUNDLE_NAME),
                        intent.getLongExtra(BUNDLE_INTERVAL, DEFAULT_BROADCAST_INTERVAL));
                broadcastThread.start();
            } else if (ACTION_STOP_BROADCAST.equals(intent.getStringExtra(BUNDLE_ACTION))) {
                if (broadcastThread != null) {
                    broadcastThread.stopBroadcast();
                }
            }
        }
        return START_STICKY;
    }

    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }
}

File: near/src/main/java/com/adroitandroid/near/discovery/client/UdpBroadcastService.kt
package com.adroitandroid.near.discovery.client

import android.app.Service
import android.content.Intent
import android.os.IBinder
import com.adroitandroid.near.discovery.server.UdpServerService
import com.adroitandroid.near.model.Host
import org.json.JSONObject

class UdpBroadcastService : Service() {
    private var broadcastThread: BroadcastThread? = null
    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
        if (ACTION_START_BROADCAST == intent.getStringExtra(BUNDLE_ACTION)) {
            val hostJsonString: String? = intent.getStringExtra(BUNDLE_HOST_JSON)
            val hostJson: JSONObject = if(hostJsonString != null) {
                JSONObject(hostJsonString)
            } else {
                JSONObject(mapOf(Host.JSON_NAME to Host.DUMMY, Host.JSON_FILTER_TEXT to ""))
            }

            broadcastThread?.stopBroadcast()
            broadcastThread = BroadcastThread(
                    hostJson,
                    intent.getLongExtra(BUNDLE_INTERVAL, DEFAULT_BROADCAST_INTERVAL),
                    intent.getIntExtra(BUNDLE_DISCOVERY_PORT, UdpServerService.DISCOVERY_PORT))
            broadcastThread!!.start()
        } else if (ACTION_STOP_BROADCAST == intent.getStringExtra(BUNDLE_ACTION)) {
            broadcastThread?.stopBroadcast()
        }
        return START_STICKY
    }

    override fun onBind(intent: Intent): IBinder? {
        return null
    }

    companion object {
        const val BUNDLE_HOST_JSON = "bundle_host_json"
        const val BUNDLE_ACTION = "bundle_action"
        const val BUNDLE_INTERVAL = "bundle_interval"
        const val BUNDLE_DISCOVERY_PORT = "bundle_discovery_port"
        const val ACTION_START_BROADCAST = "start_broadcast"
        const val ACTION_STOP_BROADCAST = "stop_broadcast"
        private const val DEFAULT_BROADCAST_INTERVAL: Long = 5000
    }
}